### ISSUE TITLE
Hash function fnv1a32 gives incorrect results.

### ISSUE THREAD

**Issue by martinvanhensbergen** [zeek/zeek#3970]
Hello team,

I believe I have found a bug in the implementation of the native fnv1a32 hash function. Using the 
function with strings as inputs, I get incorrect results.

Below I have written my analysis. It is then also followed with a usage question. Let me know how
I can help!

**--- Hash calculation**

Testscript:
```
local s: string = "foobar";

print("---");
print(s);
print(fmt("0x%x", fnv1a32(s)));
```
Output:
```
devel-vm> zeek -b ../hashtest.zeek 
---
foobar
0x65313259

```
According to two independent reference implementations, actual value should be:
**0xbf9cf968**

See:
- https://github.com/jslicer/FNV-1a/blob/master/Fnv1aTests/Fnv1a32Tests.cs
- https://md5calc.com/hash/fnv1a32/foobar

I have applied debug code as per the diff below:

```
diff --git a/src/zeek.bif b/src/zeek.bif
index eccc7f029..c3cf945f4 100644
--- a/src/zeek.bif
+++ b/src/zeek.bif
@@ -942,6 +942,8 @@ function fnv1a32%(input: any%): count
 
        for ( auto i = 0; i < desc.Len(); ++i )
                {
+               printf("Debug: %d: %c (%d)\n", i, bytes[i], bytes[i]);
+
                rval ^= (uint32_t) bytes[i];
                rval *= prime32;
                }
```

The output is now:
```
devel-vm> zeek -b ../hashtest.zeek 
---
foobar
Debug: 0:  (7)
Debug: 1:  (0)
Debug: 2:  (0)
Debug: 3:  (0)
Debug: 4:  (6)
Debug: 5:  (0)
Debug: 6:  (0)
Debug: 7:  (0)
Debug: 8: f (102)
Debug: 9: o (111)
Debug: 10: o (111)
Debug: 11: b (98)
Debug: 12: a (97)
Debug: 13: r (114)
0x65313259

```
As one can see, the hash is calculated over a representation of the string and not the value of the string itself. I believe the first 4 bytes indicate the type of the input (string) and the next 4 bytes the length of the string (followed by the string itself).

Changing the Desc to DESC_READABLE does give the correct result:

```
devel-vm> zeek -b ../hashtest.zeek 

---
foobar
Debug: 0: f (102)
Debug: 1: o (111)
Debug: 2: o (111)
Debug: 3: b (98)
Debug: 4: a (97)
Debug: 5: r (114)
0xbf9cf968

```
So I believe the `DESC_BINARY `should be changed to `DESC_READABLE `but I don't know what the effect would
be in different input types (e.g. ints, counts).

**--- Hash on bytestrings**

I would also like to calculate hashes on bytestrings instead of strings. I don't know how to use this.
I have considered the following script:

```
local s: string;

s = hexstr_to_bytestring("00010203");

print("---");
print(s);
print(|s|);
print(fmt("0x%x", fnv1a32(s)));
```

The output is:

```
devel-vm> zeek -b ../hashtest.zeek 
---
\x00\x01\x02\x03
4
Debug: 0: \ (92)
Debug: 1: x (120)
Debug: 2: 0 (48)
Debug: 3: 0 (48)
Debug: 4: \ (92)
Debug: 5: x (120)
Debug: 6: 0 (48)
Debug: 7: 1 (49)
Debug: 8: \ (92)
Debug: 9: x (120)
Debug: 10: 0 (48)
Debug: 11: 2 (50)
Debug: 12: \ (92)
Debug: 13: x (120)
Debug: 14: 0 (48)
Debug: 15: 3 (51)
0xe64618cd
```

As you can see, the length of the bytestring is correctly indicated as 4. However, feeding this into the
hash function calculates a hash over the string representation instead of the actual bytes.

How do I do this properly?

---

In short:

- Is it correct that the desc type should be DESC_READABLE or would that break other inputs?
- How do I calculate a hash over a bytestring?

Let me know if my hunch is correct and what I can do to make a PR for this.
I would like to create a PR that:

- fixes this issue
- adds the 64bit variant of the hash function
- add testvectors for this function so that it gets tested.

Kind regards,
Martin

**Comment by 0xxon** [comment]
Hi,

this is a kind of fun issue that you found here :)

From what I can tell, the reason that the fnv1a32 bif was implemented in Zeek is mostly to provide a way to round-robin values. As such, the main goal of it was to provide hash-values for any possible Zeek value. For this goal, as you found, it uses the Describe functionality of Zeek values.

The describe functionality is, essentially, meant to be able to represent the data in Zeek Values - either in binary or readable form. It is not meant to give you direct access to the underlying data. That being said, at least for strings DESC_READABLE might work out fine. For other datatypes, it probably won't really make anything worse - looking at the sourcecode, the main change for most base-types seems to be that the type information is not prepended in readable mode.

That being said, if we want this to work "correctly" (or as correctly as possible), it seems like the right way to do this would be to make the fnv1a32 bif type aware. So - for the base Values it could just hash the underlying values, and for values that it doesn't want to directly deal with (like opaque vals, etc), it could still just call the describe function.

**Comment by martinvanhensbergen** [comment]
I have added https://github.com/zeek/zeek/pull/3971

Let me know what you think :-)

**Comment by martinvanhensbergen** [comment]
I see that the testcases fail. From what I can tell (but I will investigate more later today) is that the function was indeed used in creating identifiers and ids. I think due to the values being a bit different now that some ordering of logs have changed and that an update of the baseline will do. Do you share that quick analysis or might there be something else involved?

Also I want to change the code a bit to not have the constants as parameter but separately defined first.
